### ISSUE TITLE
[ pwm ] Pare down PWM CDC

### ISSUE THREAD

**Issue by martin-lueker** [lowRISC/opentitan#6044]
This recommendation comes from @msfschaffner in PR #5556

**Comment by tjaychen** [comment]
@martin-lueker another option we are going to propose soon is a split regfile.
Ie a module can have two regfiles, each corresponding to a different domain.  There are latency concerns, so ultimately it will be up to you to decide which makes sense. 

On the parallel side though, we will probably also need to formalize a bit so we don't have too many splinter options.

**Comment by martin-lueker** [comment]
> @martin-lueker another option we are going to propose soon is a split regfile.
> Ie a module can have two regfiles, each corresponding to a different domain. There are latency concerns, so ultimately it will be up to you to decide which makes sense.
> 
> On the parallel side though, we will probably also need to formalize a bit so we don't have too many splinter options.

I'm happy to execute Michael's option 1.  However if there is an option of putting a regfile in AON then really this whole IP can go there.  There is nothing in this IP that needs to be on the bus domain, not even an interrupt.

**Comment by tjaychen** [comment]
There's nothing stopping you really from putting an entire module onto the `aon` clock domain. 
Putting the entire module there seems extreme though, but I can see it for something like pwm.

Does this mean you don't have any data or anything really you need to read back? It's pretty much configure in aon domain and let it go?

**Comment by martin-lueker** [comment]
Yeah, after seeing how @tomroberts-lowrisc did this in https://github.com/lowRISC/opentitan/pull/6895 I now agree that we should pull this whole IP into the AON clock domain.   Would somebody be about to set aside a half hour meeting to go over the process with @Muqing-Liu and me?

**Comment by tomroberts-lowrisc** [comment]
Hi @martin-lueker I can try to help. From a quick look at the pwm rtl, I think you can pretty much just copy the same approach as the AON Timer (TLUL async fifo on the incoming bus, then reg_top and everything else on the AON clock).

**Comment by tjaychen** [comment]
@tomroberts-lowrisc @martin-lueker just one thing to note, this obviously increases the access latency dramatically compared to before (i think the old approach you could basically have all the syncs happening together at the same time). 

For something like `pwm` I'm guessing it is okay because it is more or less set and forget, but I'm less certain about `aon_timer`. 
The `aon_timer` i'm less sure, because it's something that gets touched every so often, and I can also imagine software polling on it for timer reads.  We might want to be a bit careful there.  For `aon_timer` I was honestly thinking originally to split the reg (similar to `rom_ctrl`) into two portions, where some of the configuration can be on `aon`, and some of the things that might require faster response on the normal clock domain.

**Comment by tomroberts-lowrisc** [comment]
Agree this looks good for PWM - @tjaychen let's discuss AON timer in our next meeting, we have a few ideas for how we might work around the latency

**Comment by tjaychen** [comment]
hey all,
we had a sync with the software team yesterday, and in general they're
quite concerned about potentially having a lot of latency for both touching
the timers and setting up particular modules.
Specifically for the timer, it might be something that is frequently read
as part of determining the next OS tick, so there is more concern.

We are thinking about adding a primitive pulse-sync-with-data module to
handle these kinds of crossings while keeping the main regfile on the usual
clock domain.
let me know what you think.

**Comment by msfschaffner** [comment]
We now have a new autogenerated CDC option in the CSR regfile hjsons (thanks to @tjaychen). @martin-lueker can you try to use that one for PWM and then close this issue?

**Comment by tjaychen** [comment]
if it's too difficult to start (since i haven't done a good job documenting
this), l'm happy to have a quick meeting to walk through, or even help you
guys prototype it.

**Comment by Muqing-Liu** [comment]
Thank you very much for your help! I have checked the adc_ctrl.hjson file, so we need to add an "async" field for the registers, correct? I have made the corresponding changes for pwm module, and I am currently trying to simulate the module before pushing it. There are some issues with the simulation, should be resolved soon.

**Comment by msfschaffner** [comment]
@Muqing-Liu, the RTL might require some updates as well. i.e., depending on whether there was some "hand crafted" CDC already in place, that would have to be removed for this to work properly.

**Comment by Muqing-Liu** [comment]
Yes, sure. I have also updated the RTL accordingly, and the CDC was removed too. Thanks!

**Comment by tjaychen** [comment]
let us know if you run into issues as well, we can have a quick look.
There could be bugs hidden within the regfile cdc handling as well.

**Comment by Muqing-Liu** [comment]
Sure, thanks a lot!
